src/router.ts

```diff
@@ -30,29 +30,37 @@ export const IS_PATTERN = /[*:{}+?()]/;
 
 export class UrlPatternRouter<T> implements Router<T> {
   readonly _routes: Route<T>[] = [];
-  readonly _middlewares: T[] = [];
+  readonly _middlewares: Route<T>[] = [];
 
-  addMiddleware(fn: T): void {
-    this._middlewares.push(fn);
+  isURLPattern(value: unknown): value is URLPattern {
+    return value instanceof URLPattern;
   }
 
-  add(method: Method | "ALL", pathname: string | URLPattern, handlers: T[]) {
-    if (
-      typeof pathname === "string" && pathname !== "/*" &&
-      IS_PATTERN.test(pathname)
-    ) {
-      this._routes.push({
-        path: new URLPattern({ pathname }),
-        handlers,
-        method,
-      });
-    } else {
-      this._routes.push({
-        path: pathname,
-        handlers,
-        method,
-      });
-    }
+  addMiddleware(pathname: string | URLPattern, handler: T): void {
+    this._middlewares.push({
+      path: this.isURLPattern(pathname)
+        ? pathname
+        : new URLPattern({ pathname }),
+      handlers: [handler],
+      method: "ALL",
+    });
+  }
+
+  add(
+    method: Method | "ALL",
+    pathname: string | URLPattern,
+    handlers: T[],
+  ): void {
+    this._routes.push({
+      path: (
+          typeof pathname === "string" && pathname !== "/*" &&
+          !this.isURLPattern(pathname)
+        )
```

Comment:
Nit: To clarify, my idea was to have everything inside this condition be factored out. Not just the `URLPattern` bit. Also, I don't see a reason why it should be a method when it doesn't use any property from this class.